This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [616]:
# get api key from your .env file
import os
from dotenv import load_dotenv

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')

print(API_KEY)

zi8z9mYwCJWJQjgsLyed


Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [617]:
# First, import the relevant modules
import requests
import collections
import json

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [618]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url = "https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X"
params = {"start_date": "2017-01-01", "end_date": "2017-01-01", "api_key": API_KEY}
r = requests.get(url, params=params)
data_ = json.loads(r.text)

In [619]:
print(type(data_))

<class 'dict'>


In [620]:
print(data_)

{'dataset': {'id': 10095370, 'dataset_code': 'AFX_X', 'database_code': 'FSE', 'name': 'Carl Zeiss Meditec (AFX_X)', 'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704', 'refreshed_at': '2020-12-01T14:48:09.907Z', 'newest_available_date': '2020-12-01', 'oldest_available_date': '2000-06-07', 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'frequency': 'daily', 'type': 'Time Series', 'premium': False, 'limit': None, 'transform': None, 'column_index': None, 'start_date': '2017-01-01', 'end_date': '2017-01-01', 'data': [], 'collapse': None, 'order': None, 'database_id': 6129}}


In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

{'dataset': {'id': 10095370,
  'dataset_code': 'AFX_X',
  'database_code': 'FSE',
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'refreshed_at': '2020-12-01T14:48:09.907Z',
  'newest_available_date': '2020-12-01',
  'oldest_available_date': '2000-06-07',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2021-01-03',
  'end_date': '2020-12-01',
  'data': [],
  'collapse': None,
  'order': None,
  'database_id': 6129}}

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [621]:
url = "https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X"
params = {"start_date": "2017-01-01", "end_date": "2017-12-31", "api_key": API_KEY}
r = requests.get(url, params=params)
data = json.loads(r.text)
print(type(data))

<class 'dict'>


In [622]:
dataset = data['dataset']
column_names = dataset['column_names']

In [623]:
data_lists = data['dataset']['data']

In [624]:
dates_all = {}
dates_opens = {}
opens = []
highs = []
lows = []
changes = []
traded_volumes = []
dates_highs = {}
dates_lows = {}
dates_changes = {}
x = range(12)
for list_ in data_lists:
    dates_all[list_[0]]= {'open': list_[1], 'high': list_[2], 'low': list_[3], 'close': list_[4], 'change': list_[5], 'Traded Volume': list_[6], 'Turnover': list_[7], 'Last Price of the Day': list_[8], 'Daily Traded Units': list_[9], 'Daily Turnover': list_[10]}
    dates_opens[list_[0]] =  list_[1]
    dates_highs[list_[0]]= list_[2]
    dates_lows[list_[0]]= list_[3]
    dates_changes[list_[0]]= list_[5]
    if list_[1] != None:
        opens.append(list_[1])
    if list_[5] != None:
        changes.append(list_[5])
    if list_[2] != None:
        highs.append(list_[2])
    if list_[3] != None:
        lows.append(list_[3])
    if list_[6] != None:
        traded_volumes.append(list_[6])
            

In [625]:
highest_open = max(opens)
lowest_open = min(opens)
print(highest_open)
print(lowest_open)

53.11
34.0


In [626]:
max_change_high_low = highest_open - lowest_open
print(max_change_high_low)

19.11


In [627]:
max_change_closing_price = max(changes)
print(max_change_closing_price)

-0.44


In [628]:
avg_traded_volume = sum(traded_volumes) / len(traded_volumes)
print(avg_traded_volume)

89124.33725490196


In [629]:
def calc_median(x):
    s = sorted(x)
    n = len(x)
    half = len(s)//2
    if n % 2 == 0:
        median= (s[half] + s[(half - 1)])/2
        return median
    else:
        median2 = s[half]
        return median2

In [630]:
y = [0, 0, 0, 0, 4, 4, 6, 8]
j = [1, 1, 2, 4]
k = [0, -5, 2, 8, 6, -9, 9, 10]
calc_median(k)

4.0

In [631]:
calc_median(traded_volumes)

76286.0

3. Calculate what the highest and lowest opening prices were for the stock in this period.

        The highest opening price for 2017 was 53.11. The lowest opening price was 34.0.
    
    
4. What was the largest change in any one day (based on High and Low price)?

        The largest change based on High and Low price is 19.11.
    
    
5. What was the largest change between any two days (based on Closing Price)?

        The largest change based on Closing Price was -0.44. A majority of these values were None.
    
    
6. What was the average daily trading volume during this year?

        Average daily trading volume for 2017 was 89124.34.
    
    
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

        Median trading volume during 2017 was 76286.0.